internal/cmd/generate.go

```diff
@@ -57,66 +63,58 @@ func Generate(dir string, stderr io.Writer) (map[string]string, error) {
 	output := map[string]string{}
 	errored := false
 
+	var pairs []outPair
 	for _, sql := range conf.SQL {
-		combo := config.Combine(conf, sql)
-		name := combo.Go.Package
+		if sql.Gen.Go != nil {
+			pairs = append(pairs, outPair{
+				SQL: sql,
+				Gen: config.SQLGen{Go: sql.Gen.Go},
+			})
+		}
+		if sql.Gen.Kotlin != nil {
+			pairs = append(pairs, outPair{
+				SQL: sql,
+				Gen: config.SQLGen{Kotlin: sql.Gen.Kotlin},
+			})
+		}
+	}
+
+	for _, sql := range pairs {
+		combo := config.Combine(conf, sql.SQL)
 		var result dinosql.Generateable
 
 		// TODO: This feels like a hack that will bite us later
 		sql.Schema = filepath.Join(dir, sql.Schema)
 		sql.Queries = filepath.Join(dir, sql.Queries)
 
-		switch sql.Engine {
-		case config.EngineMySQL:
-			// Experimental MySQL support
-			q, err := mysql.GeneratePkg(name, sql.Schema, sql.Queries, combo)
-			if err != nil {
-				fmt.Fprintf(stderr, "# package %s\n", name)
-				if parserErr, ok := err.(*dinosql.ParserErr); ok {
-					for _, fileErr := range parserErr.Errs {
-						printFileErr(stderr, dir, fileErr)
-					}
-				} else {
-					fmt.Fprintf(stderr, "error parsing schema: %s\n", err)
-				}
-				errored = true
-				continue
-			}
-			result = q
-
-		case config.EnginePostgreSQL:
-			c, err := dinosql.ParseCatalog(sql.Schema)
-			if err != nil {
-				fmt.Fprintf(stderr, "# package %s\n", name)
-				if parserErr, ok := err.(*dinosql.ParserErr); ok {
-					for _, fileErr := range parserErr.Errs {
-						printFileErr(stderr, dir, fileErr)
-					}
-				} else {
-					fmt.Fprintf(stderr, "error parsing schema: %s\n", err)
-				}
-				errored = true
-				continue
-			}
-
-			q, err := dinosql.ParseQueries(c, sql)
-			if err != nil {
-				fmt.Fprintf(stderr, "# package %s\n", name)
-				if parserErr, ok := err.(*dinosql.ParserErr); ok {
-					for _, fileErr := range parserErr.Errs {
-						printFileErr(stderr, dir, fileErr)
-					}
-				} else {
-					fmt.Fprintf(stderr, "error parsing queries: %s\n", err)
-				}
-				errored = true
-				continue
-			}
-			result = q
+		var name string
+		parseOpts := dinosql.ParserOpts{}
+		if sql.Gen.Go != nil {
+			name = combo.Go.Package
+		} else if sql.Gen.Kotlin != nil {
+			parseOpts.UsePositionalParameters = true
+			name = combo.Kotlin.Package
+		}
 
+		result, errored = parse(name, dir, sql.SQL, combo, parseOpts, stderr)
+		if errored {
+			break
 		}
 
-		files, err := dinosql.Generate(result, combo)
+		var files map[string]string
+		var out string
+		if sql.Gen.Go != nil {
```

Comment:
The configuration format supports multiple language packages per package. This if statement means that only Go or Kotlin will be output. Code for both languages should be able to be output in a single pass.

Reply:
I believe this is correct. See lines 67-80 where I permute (engine, language) into the `outPair` struct. The parser needs to be run once per language because kotlin needs to run the parser with different options. I split out the parser changes to https://github.com/kyleconroy/sqlc/pull/306 earlier if you want to look closer there (though it's squashed into this PR and is outdated)

Reply:
Yep, I missed the fact that you were creating multiple output pairs. Looks good!